### PR TITLE
reconcile-cluster-role-bindings followup

### PR DESCRIPTION
**Pull request by liggitt**
Add diagnostic for role bindings
Updates compatibility test to reconcile rolebindings

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k PTAL

**Comment by liggitt** [comment]
@sosiouxme, PTAL

File: pkg/diagnostics/cluster/rolebindings.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+const (
+	ClusterRoleBindingsName = "ClusterRoleBindings"
+)
+
+func (d *ClusterRoleBindings) Name() string {
+	return ClusterRoleBindingsName
+}
+
+func (d *ClusterRoleBindings) Description() string {
+	return "Check that the ClusterRoleBindings are up-to-date"
```
Comment (sosiouxme) [inline pkg/diagnostics/cluster/rolebindings.go]: Could this be slightly more descriptive for admins who are policy ignoramuses? Might be the first time they ever hear of it.

Maybe just "Check if the ClusterRoleBindings need updating for the current cluster version"

File: pkg/diagnostics/cluster/rolebindings.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		if err != nil {
+			r.Error("CRBD1002", err, fmt.Sprintf("Unable to get clusterrolebinding/%s: %v", changedClusterRoleBinding.Name, err))
+		}
+
+		missingSubjects, extraSubjects := policycmd.Diff(changedClusterRoleBinding.Subjects, actualClusterRole.Subjects)
+		switch {
+		case len(missingSubjects) > 0:
+			// Only a warning, because they can remove things like self-provisioner role from system:unauthenticated, and it's not an error
+			r.Warn("CRBD1003", nil, fmt.Sprintf("clusterrolebinding/%s is missing expected subjects.\n\nUse the `oadm policy reconcile-cluster-role bindings` command to update the role binding to include expected subjects.", changedClusterRoleBinding.Name))
```
Comment (sosiouxme) [inline pkg/diagnostics/cluster/rolebindings.go]: Shouldn't it be reconcile-cluster-role-bindings? (missing hyphen)

File: pkg/diagnostics/cluster/rolebindings.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		case len(extraSubjects) > 0:
+			// Only info, because it is normal to use policy to grant cluster roles to users
+			r.Info("CRBD1004", fmt.Sprintf("clusterrolebinding/%s has more subjects than expected.\n\nUse the `oadm policy reconcile-cluster-role bindings` command to update the role binding to remove extra subjects.", changedClusterRoleBinding.Name))
```
Comment (sosiouxme) [inline pkg/diagnostics/cluster/rolebindings.go]: same here

**Comment by sosiouxme** [comment]
@liggitt LGTM otherwise.

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go [hunk 4]
```diff
@@ -284,11 +284,11 @@ func contains(list []kapi.ObjectReference, item kapi.ObjectReference) bool {
 	return false
 }
 
-// diff returns lists containing the items unique to each provided list:
+// Diff returns lists containing the items unique to each provided list:
 //   list1Only = list1 - list2
 //   list2Only = list2 - list1
 // if both returned lists are empty, the provided lists are equal
-func diff(list1 []kapi.ObjectReference, list2 []kapi.ObjectReference) (list1Only []kapi.ObjectReference, list2Only []kapi.ObjectReference) {
+func Diff(list1 []kapi.ObjectReference, list2 []kapi.ObjectReference) (list1Only []kapi.ObjectReference, list2Only []kapi.ObjectReference) {
```
Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: If you expose this, it needs a better name.

File: pkg/diagnostics/cluster/rolebindings.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	reconcileOptions := &policycmd.ReconcileClusterRoleBindingsOptions{
+		Confirmed:         false,
+		Union:             false,
+		Out:               ioutil.Discard,
+		RoleBindingClient: d.ClusterRoleBindingsClient.ClusterRoleBindings(),
+	}
+
+	changedClusterRoleBindings, err := reconcileOptions.ChangedClusterRoleBindings()
+	if err != nil {
+		r.Error("CRBD1000", err, fmt.Sprintf("Error inspecting ClusterRoleBindings: %v", err))
```
Comment (deads2k) [inline pkg/diagnostics/cluster/rolebindings.go]: Why not return

File: pkg/diagnostics/cluster/rolebindings.go (hunk 2 above)
Comment (deads2k) [inline pkg/diagnostics/cluster/rolebindings.go]: Display which subjects are missing.

Comment (deads2k) [inline pkg/diagnostics/cluster/rolebindings.go]: > Display which subjects are missing.

Oh, I started it didn't I?  I guess the division of levels makes sense.

File: pkg/diagnostics/cluster/rolebindings.go (hunk 5 above)
Comment (deads2k) [inline pkg/diagnostics/cluster/rolebindings.go]: > Why not return

This was originally me too.  You should be careful who's code you trust as a template :)

**Comment by liggitt** [comment]
comments addressed

**Comment by liggitt** [comment]
[merge]

**Pull request merged**